src/middlewares/Cache.ts

```diff
@@ -0,0 +1,155 @@
+     * else let the next middleware in pileline be invoked and cache it.
+     *
+     * @param  {FetchRequest} options
+     * @param  {callback}       next
+     * @return {any}
+     */
+    public process(options: FetchRequest, next: (nextOptions: FetchRequest) => Promise<FetchResponse<any>>): any {
+        const bucket = this.getEffectiveBucket(options);
+        const cacheable: boolean = (options.method || "GET").toLowerCase() === "get";
+
+        return cacheable && bucket && this.getCachedResponse(options, bucket)
+        || next(options).then((response: FetchResponse<any>) => {
+            if (!cacheable || !bucket) { return response; }
+
+            const clone = response.clone();
+            const ttl = bucket.options.maxAge || this.config.maxAge || 600; // TTL fallback to 10 min
+            clone.text().then((text: string) => this.storage.setItem(this.getCacheKey(options, bucket), text, ttl) );
+
+            return response;
+        });
+    }
+
+    /**
+     * If possible return the desired type of store or fallback to default in-memory store.
+     *
+     * @return {ICacheMiddlewareStore}
+     */
+    protected getDefaultCacheStorage(): ICacheMiddlewareStore {
+        if (this.config.storage === "localStorage") {
+            return new LocalStorageCache();
+        }
+
+        if (this.config.storage === "redis") {
+            return new RedisCache();
+        }
+
+        return new MemoryCache();
+    }
+
+    /**
+     * Returns a Promise that resolves to Response object using cached value. Returns null if the cache is non existent or expired.
+     *
+     * @param  {FetchRequest} options
+     * @param  {ICacheBucket} bucket
+     * @return {Promise|null}
+     */
+    protected getCachedResponse(options: FetchRequest, bucket: ICacheBucket): Promise<FetchResponse<any>>|null {
+        const value = this.storage.getItem(this.getCacheKey(options, bucket));
+        if (!value) { return null; }
+
+        return Promise.resolve(new Response(value, undefined));
+    }
+
+    /**
+     * Returns a bucket that is active and can actually respond to currently requested endpoint.
+     *
+     * @param  {FetchRequest} options
+     * @return {ICacheBucket|null}
+     */
+    protected getEffectiveBucket(options: FetchRequest): ICacheBucket|null {
+        return this.buckets.find((bucket: ICacheBucket) => {
+            return bucket.options.enable === true
+                && options.url !== undefined
+                && bucket.options.path === options.url;
+        });
+    }
+
+    /**
+     * Gets the normalized cache key which is unique to request with respect to uri and params.
+     *
+     * @param  {FetchRequest} options
+     * @param  {ICacheBucket} bucket
+     * @return {string}
+     */
+    protected getCacheKey(options: FetchRequest, bucket: ICacheBucket): string {
+        let key: string = bucket.options.key || options.url || bucket.options.path;
```

Comment:
it can be two. `bucket.options.key || bucket.options.path`, as key is optional in bucket. `options.url` is not needed here.